src/viper/README.md

```diff
@@ -240,33 +513,53 @@ Below, we summarize the language features that _Motoko-san_ currently supports.
     * `assert:func` — Function preconditions
     
     * `assert:return` — Function postconditions. 
-    
-        * These may refer to variables in the _initial_ state of the function call using the syntax `(old <exp>)`, for example:
-
-            ```motoko
-            var x : Int;
-            private func dec() : () {
-                x -= 1;
-                assert:return x < old(x);
-            };
-            ```
 
-            is equivalent to
-
-            ```motoko
-            var x : Int;
-            private func dec() : () {
-                let old_x = x;
-                x -= 1;
-                assert:return x < old_x;
-            };
-            ```
-    
     * `assert:system` — Compile-time assertions
     
-    **Loop invariants** — Extension is _simple_.
+    * `assert:loop:invariant` — Loop invariants
     
-    **Pure functions** — The tool could be easily extended with a keyword, e.g., `@pure`, to specify functions that are verifier to be side-effect free; such functions could be used inside other code specifications, e.g., `assert:invariant is_okay()` for some `@pure func is_okay() : Bool`.
+    **Expressions in assertions** — Note that using the syntax bellow outside of assertions causes undefined behavior.
+
+    * `e1 implies e2` refers to logical implication.
+
+    * `old(e)` — refers to value of the expression `e` in the _initial_ state of the function call. Could be used in postconditions or loop invariants to verify changes that are made. For example:
+
+      ```motoko
+      var x : Int;
+      private func dec() : () {
+          x -= 1;
+      assert:return x < old(x);
+      };
+      ```
+
+      is equivalent to
+
+      ```motoko
+      var x : Int;
+      private func dec() : () {
+          let old_x = x;
+          x -= 1;
+          assert:return x < old_x;
+      };
+      ```
+
+    * `Prim.Forall<T>(func x ..)` and `Prim.Exists<T>(func x ..)` — refers ∀ and ∃ quantifiers. For example:
+
+      ```motoko
+      private func fill_array(x: [var Bool], val: Bool): () {
+         assert:return Prim.Forall<Nat>(func i = (0 <= i and i < x.size() implies a[i] == val));
+      }
+      ```
+    * `Prim.Ret<T>()` — refers to the result value of a function. It's temporal syntax and requires to manually annotate type of the value. Example:
```

Comment:
```suggestion
    * `Prim.Ret<T>()` — refers to the result value of a function. It's temporary syntax and requires to manually annotate the type of the value. Example:
```

Actually, should it be Prim.ret (lowercase) to match Prim.forall and Prim.exists (as implemented).